### PR TITLE
Recursively convert (adapt) `Semidiscretization` object to GPU types with Adapt.jl

### PR DESCRIPTION
**Pull request by efaulhaber**
Second step of #484.
Based on #486.

### CODE REVIEW

File: src/general/gpu.jl [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+# the `CuArray` type.
+# What this does is that it converts all `Array`s inside this system to `CuArray`s,
+# therefore copying them to the GPU.
+# In order to run a simulation on a GPU, we want to call `adapt(T, semi)` to adapt the
+# `Semidiscretization` `semi` to the GPU array type `T` (e.g. `CuArray`).
+#
+# `Adapt.@adapt_structure` automatically generates the `adapt` function for our custom types.
+Adapt.@adapt_structure Semidiscretization
+Adapt.@adapt_structure WeaklyCompressibleSPHSystem
+Adapt.@adapt_structure DensityDiffusionAntuono
```
Comment (svchb) [inline src/general/gpu.jl:12]: Add here BoundaryModelMonaghanKajtar!

File: src/general/semidiscretization.jl [hunk 2]
```diff
@@ -248,7 +248,7 @@ timespan: (0.0, 1.0)
 u0: ([...], [...]) *this line is ignored by filter*
 ```
 """
-function semidiscretize(semi, tspan; reset_threads=true)
+function semidiscretize(semi, tspan; reset_threads=true, data_type=nothing)
```
Comment (svchb) [inline src/general/semidiscretization.jl:251]: fix doc string

File: src/general/semidiscretization.jl [hunk 3]
```diff
@@ -274,8 +274,14 @@ function semidiscretize(semi, tspan; reset_threads=true)
 
     sizes_u = (u_nvariables(system) * n_moving_particles(system) for system in systems)
     sizes_v = (v_nvariables(system) * n_moving_particles(system) for system in systems)
-    u0_ode = Vector{ELTYPE}(undef, sum(sizes_u))
-    v0_ode = Vector{ELTYPE}(undef, sum(sizes_v))
+
+    if isnothing(data_type)
```
Comment (svchb) [inline src/general/semidiscretization.jl:278]: add comment

File: src/general/semidiscretization.jl [hunk 4]
```diff
@@ -286,6 +292,15 @@ function semidiscretize(semi, tspan; reset_threads=true)
         write_v0!(v0_system, system)
     end
 
+    if !isnothing(data_type)
+        # Convert all arrays to the correct array type. When e.g. `data_type=CuArray`, this will
+        # convert all `Array`s to `CuArray`s, moving data to the GPU.
+        # See the comments in general/gpu.jl for more details.
+        semi_adapted = Adapt.adapt(data_type, semi)
+
+        return DynamicalODEProblem(kick!, drift!, v0_ode, u0_ode, tspan, semi_adapted)
```
Comment (svchb) [inline src/general/semidiscretization.jl:301]: can you not just overwrite `semi`?

**Review by svchb — CHANGES_REQUESTED** [review]

File: src/general/semidiscretization.jl (hunk 4 above)
Comment (efaulhaber) [reply src/general/semidiscretization.jl:301]: No, `Semidiscretization` is not mutable, and cannot be mutable to be GPU-compatible.

File: src/general/semidiscretization.jl (hunk 2 above)
Comment (efaulhaber) [reply src/general/semidiscretization.jl:251]: I'll add this in #493 when this feature is actually working.

**Review by svchb — APPROVED** [review]

**Pull request merged**